Mi 1221 1

manojmalik20 opened the pull request:
<!-- Thank you for contributing a pull request! Here are a few tips to help you:

1. If this is your first contribution, make sure you've read the Contribution Checklist https://developers.mattermost.com/contribute/getting-started/contribution-checklist/
2. Read our blog post about "Submitting Great PRs" https://developers.mattermost.com/blog/2019-01-24-submitting-great-prs
3. Take a look at other repository specific documentation at https://developers.mattermost.com/contribute
-->

#### Summary
<!--
A description of what this pull request does.
-->

#### Ticket Link
<!--
If this pull request addresses a Help Wanted ticket, please link the relevant GitHub issue, e.g.

  Fixes https://github.com/mattermost/mattermost-server/issues/XXXXX

Otherwise, link the JIRA ticket.
-->

#### Related Pull Requests
<!--
List all PRs related to resolving a ticket. For instance, if you submitted a PR to `mattermost/mattermost-server`, please include it here.
-->
- Has server changes (please link here)
- Has redux changes (please link here)
- Has mobile changes (please link here)

#### Screenshots
<!--
If the PR includes UI changes, include screenshots/GIFs.
-->

manojmalik20 committed:
Review fixes and added timepicker menu
Changed the custom status selector to a simple function and changed its use everywhere
Made constants for the expiry menu items and used them
Changed the expiry menu items to use localized text
Removed the daypicker.scss file and did the changes in _daypicker.scss file
Added all the localization ids in en.json

manojmalik20 committed:
Completed the timepicker menu
Made a util for getting current date and time for timezone
Made the function to round current time to nearest 30 min time
Made the function to fill the menu with 30 minute intervals
Unified the separate state variables for day and time to only time
Made the dateTimeInput component to be controlled by the custom status modal
Fixed the styles for the timepicker and daypicker
Fixed lint errors and type check errors

manojmalik20 committed:
Fixed some issues in daypicker
Fixed the styling in daypicker
Added logic for disabled days in daypicker
Fixed the issue of prevMonth button not working in daypicker

chetanyakan commented on selectors/views/custom_status.ts:
```diff
@@ -12,17 +12,15 @@ import {UserCustomStatus} from 'mattermost-redux/types/users';
-        return rv;
+    // if (Date.now() < timer) {
+    return rv;
 
-        // }
-    };
+    // }
```
Directly return from line 18 instead of creating a variable.

manojmalik20 committed:
Complete expiry support in custom status modal
Changed the custom status type to include duration and expires_at and made enum duration
Removed the duration constants from constants file and used enum type everywhere
Added localization in the expiry menu values
Modified custom status modal to send expiry with the custom status
Modified custom status modal to handle initial custom expiry time
Modified the disable set status variable to handle the confirm button disability
Modified the custom status selector to return custom status only when it's not expired

chetanyakan commented on components/custom_status/custom_status_modal.tsx:
```diff
@@ -49,20 +54,43 @@ const defaultCustomStatusSuggestions: DefaultUserCustomStatus[] = [
+    const userTimezone = useSelector((state: GlobalState) => getUserTimezone(state, currentUserId));
+    const enableTimezone = useSelector(areTimezonesEnabledAndSupported);
+
+    let currentTime: Date;
+    let timezone: string | undefined;
+    if (enableTimezone) {
+        if (userTimezone.useAutomaticTimezone) {
+            currentTime = getCurrentDateAndTimeForTimezone(userTimezone.automaticTimezone);
+            timezone = userTimezone.automaticTimezone;
+        } else {
+            currentTime = getCurrentDateAndTimeForTimezone(userTimezone.manualTimezone);
+            timezone = userTimezone.manualTimezone;
+        }
+    } else {
+        currentTime = new Date();
+    }
```
I think this can still be improved, but here's a suggestion:

```suggestion
    let currentTime = new Date();
    let timezone: string | undefined;
    if (enableTimezone) {
        timezone = userTimezone.manualTimezone;
        if (userTimezone.useAutomaticTimezone) {
            timezone = userTimezone.automaticTimezone;
        }
        currentTime = getCurrentDateAndTimeForTimezone(timezone);
    }
```

chetanyakan commented on components/custom_status/date_time_input.tsx:
```diff
@@ -61,58 +57,87 @@ const Navbar: React.FC<Partial<NavbarElementProps>> = (navbarProps: Partial<Navb
+    const setTimeAndOptions = () => {
+        const currentTime = timezone ? getCurrentDateAndTimeForTimezone(timezone) : new Date();
+        let startTime: Date;
+        if (time.getDate() === currentTime.getDate()) {
+            startTime = getRoundedTime(currentTime);
         } else {
-            currentDate = getCurrentDateForTimezone(userTimezone.manualTimezone);
+            startTime = moment(time).startOf('day').toDate();
         }
```
Initialise the value in the let statement and ditch the else

chetanyakan replied:
Initialize the value in the let statement and ditch the else

chetanyakan commented on components/custom_status/date_time_input.tsx:
```diff
@@ -61,58 +57,87 @@ const Navbar: React.FC<Partial<NavbarElementProps>> = (navbarProps: Partial<Navb
-    } else {
-        currentDate = new Date();
-    }
+        setTimeOptions(getDateInIntervals(startTime, CUSTOM_STATUS_TIME_PICKER_INTERVALS_IN_MINUTES));
```
You are passing the value `CUSTOM_STATUS_TIME_PICKER_INTERVALS_IN_MINUTES` as a parameter in `getDateInIntervals`, but using it directly from constants in `getRoundedTime`.

Why is this behavior inconsistent?

chetanyakan commented on components/custom_status/expiry_menu.tsx:
```diff
@@ -80,17 +83,17 @@ const ExpiryMenu: React.FC<Props> = (props: Props) => {
                         <Menu.Group>
                             {Object.keys(expiryMenuItems).map((item, index) => (
                                 <Menu.ItemAction
                                     key={index}
```
Don't use `index` as react keys.
We can use `item` if it is unique.

manojmalik20 committed:
Review fixes
Code refactoring in custom status modal and dateTime input and expiry menu

chetanyakan commented on components/custom_status/custom_status_modal.tsx:
```diff
@@ -4,26 +4,31 @@ import React, {useEffect, useRef, useState} from 'react';
 import {setCustomStatusInitialisationState} from 'mattermost-redux/actions/preferences';
 import {Preferences} from 'mattermost-redux/constants';
-import {UserCustomStatus} from 'mattermost-redux/types/users';
+import {UserCustomStatus, Duration} from 'mattermost-redux/types/users';
```
Should we rename the type to `CustomStatusDuration`

manojmalik20 committed:
Changed name of enum Duration to CustomStatusDuration
